features/fout-meldingen.feature

```diff
@@ -0,0 +1,168 @@
+#language: nl
+Functionaliteit: Gebruiksvriendelijke foutmeldingen
+
+    Als developer van een consumer applicatie
+    Wil ik gebruiksvriendelijke en duidelijke foutmeldingen
+    Zodat ik deze één op één kan gebruiken in de UI van mijn applicatie
+
+Rule: minimaal één optionele parameter is opgegeven
+
+    Scenario: Er zijn geen parameters opgegeven
+        Als '/wozobjecten' wordt aangeroepen
+        Dan bevat de response de volgende kenmerken
+        | naam   | waarde                                                                                                                                                                         |
+        | title  | Tenminste één parameter moet worden opgegeven.                                                                                                                                 |
+        | status | 400                                                                                                                                                                            |
+        | detail | Er moet minimaal één van de parameters 'rsin', 'kvkNummer', 'adresseerbaarObjectIdentificatie', 'nummeraanduidingIdentificatie' of 'postcode' met 'huisnummer' worden opgegeven |
+        En bevat de response geen invalidParams
+
+Rule: alleen gespecificeerde parameters mogen worden opgegeven
+
+    Abstract Scenario: Niet gespecificeerd parameter
+        Als '/wozobjecten?<parameter>=123456789' wordt aangeroepen
+        Dan bevat de response de volgende kenmerken
+        | naam   | waarde                                                     |
+        | title  | Een of meerdere parameters zijn niet correct.              |
+        | status | 400                                                        |
+        | detail | Zoeken met parameter '<parameter>' wordt niet ondersteund. |
+        En bevat de response de volgende invalidParams
+        | name        | reason                 |
+        | <parameter> | wordt niet ondersteund |
```

Comment:
De invalidParams reason waarden in foutafhandeling.feature zijn allemaal zinnen die beginnen met een hoofdletter en eindigen met een punt.

Reply:
In deze [foutafhandeling feature](https://github.com/VNG-Realisatie/Haal-Centraal-common/blob/master/features/foutafhandeling.feature) staat deze situatie al beschreven:
| parameters       | Parameter is niet verwacht.                               | unknownParam |

niet handig om dat dan hier ook op te nemen. Dit kan denk ik generiek zijn voor de HC API's dus in de common feature.

Reply:
Tja. De feature file beschrijft van het gedrag van de API voor iedere stakeholder, omdat de common zoals hij nu is vragen oproept bij developers van BAG en WOZ. Blijkbaar omdat deze niet duidelijk genoeg is. Ik heb aan Melvin gevraagd om te beschrijven wat hij graag zou willen zien als consumer. Het gaat toch om de DX van consumer en duidelijkheid over het gewenste gedrag van de API voor provider developer (die daar om vraagt). Gaan deze belangen niet boven het hebben van generieke features in de Haal Centraal Comon? Dit is denk ik de principiële vraag die we als team moeten beantwoorden. Graag reactie!

Reply:
De belangrijkste afweging in mijn beleving is hoe gedetailleerd wil je de foutsituaties terugkoppelen aan de consumers. Het optimaliseren van deze meldingen om bijvoorbeeld direct door te koppen naar de eindgebruiker betekent dat je behoorlijk fijnmazige foutmeldingen krijgt. Als je die ook nog eens van te voren wil ontdekken en specificeren zit daar een prijs aan. Hoe fijnmaziger je dit doet, hoe kleiner de kans is dat je het generiek kan definiëren. Het is dus in mijn beleving een kosten / sourcing vraagstuk. Heb je het er voor over om dit voor elke API zo fijnmazig te doen.
Rechtvaardigt het (potentiele) gebruik het ook om dat te doen.